server/public/shared/markdown/emoji.go

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package markdown
+
+import (
+	"regexp"
+)
+
+// Based off the mobile app's emoji parsing from https://github.com/mattermost/commonmark.js
+
+var (
+	nonWordRegex = regexp.MustCompile(`^\W+$`)
+	emojiRegex   = regexp.MustCompile(`^:([a-z0-9_\-+]+):\B`)
+)
+
+// parseEmoji attempts to parse a named emoji (eg. :taco:) starting at the current parser position. If an emoji is
+// found, it adds that to p.inlines and returns true. Otherwise, it returns false.
+func (p *inlineParser) parseEmoji() bool {
+	// Only allow emojis after non-word characters
+	if p.position > 1 {
+		prevChar := p.raw[p.position-1]
+
+		if isWordByte(prevChar) {
+			return false
+		}
+	}
+
+	remaining := p.raw[p.position:]
+
+	loc := emojiRegex.FindStringIndex(remaining)
+	if loc == nil {
+		return false
+	}
+
+	// Note that there may not be a system or custom emoji that exists with this name
```

Comment:
This matches the mobile app's emoji rendering where we don't actually know if an emoji exists until after this point, so anything that looks like an emoji will be able to mention anything.

If we decide to actually use this logic to show emojis in email notifications (which could be a good HW ticket after this), the renderer would have to be made aware of which emojis do and do not exist on the server.